Packet++/header/LdapLayer.h

```diff
@@ -266,10 +288,295 @@ namespace pcpp
+		 * @param[in] filterRecord The filter for the search. Please note that parsing for the search filter
+		 * doesn't exist yet. Therefore, the expected input value should be a plain ASN.1 record
+		 * @param[in] attributes A set of attributes to request for inclusion in entries that match the search
+		 * criteria and are returned to the client
+		 * @param[in] controls A vector of LDAP controls. This is an optional parameter, if not provided the message
+		 * will be created without LDAP controls
+		 */
+		LdapSearchRequestLayer(
+			uint16_t messageId, const std::string& baseObject, SearchRequestScope scope, DerefAliases derefAliases,
+			uint8_t sizeLimit, uint8_t timeLimit, bool typesOnly, Asn1Record* filterRecord,
+			const std::vector<std::string>& attributes, const std::vector<LdapControl>& controls = std::vector<LdapControl>());
+
+		/**
+		 * @return The base object for the LDAP search request entry
+		 */
+		std::string getBaseObject() const;
+
+		/**
+		 * @return The portion of the target subtree that should be considered
+		 */
+		SearchRequestScope getScope() const;
+
+		/**
+		 * @return The alias dereferencing behavior
+		 */
+		DerefAliases getDerefAlias() const;
+
+		/**
+		 * @return The maximum number of entries that should be returned from the search
+		 */
+		uint8_t getSizeLimit() const;
+
+		/**
+		 * @return The time limit for the search in seconds
+		 */
+		uint8_t getTimeLimit() const;
+
+		/**
+		 * @return If this flag is true, then it indicates that entries that match the search criteria should be
+		 * returned containing only the attribute descriptions for the attributes contained in that entry but
+		 * should not include the values for those attributes. If this flag is false, then it indicates that the
+		 * attribute values should be included in the entries that are returned
+		 */
+		bool getTypesOnly() const;
+
+		/**
+		 * @return The filter for the search. Please note that parsing for the search filter doesn't exist yet.
+		 * Therefore, the return value is a plain ASN.1 record
+		 */
+		Asn1Record* getFilter() const;
+
+		/**
+		 * @return A list of search request attributes
+		 */
+		std::vector<std::string> getAttributes() const;
+
+		template <typename Method, typename ResultType>
+		bool tryGet(Method method, ResultType& result)
+		{
+			return internalTryGet(this, method, result);
+		}
+
+	protected:
+		friend LdapLayer* LdapLayer::parseLdapMessage(uint8_t* data, size_t dataLen, Layer* prevLayer, Packet* packet);
+
+		static constexpr int baseObjectIndex = 0;
+		static constexpr int scopeIndex = 1;
+		static constexpr int derefAliasIndex = 2;
+		static constexpr int sizeLimitIndex = 3;
+		static constexpr int timeLimitIndex = 4;
+		static constexpr int typesOnlyIndex = 5;
+		static constexpr int filterIndex = 6;
+		static constexpr int attributesIndex = 7;
+
+		LdapSearchRequestLayer(std::unique_ptr<Asn1Record>& asn1Record, uint8_t* data, size_t dataLen, Layer* prevLayer, Packet* packet)
```

Comment:
Did not notice that `LdapLayer` had a similar constructor on the last PR, so this comment is regarding that class's ctor too.

Generally do **NOT** pass `unique_ptr` by non-`const` reference, pass it by value (and use moves if necessary). 

Passing it as a non-`const` reference introduces ambiguity, as it symbolizes that the smart pointer *might* be taken ownership of, which is not the case for most functions. 
**NB:** If the function actually does *conditionally* take ownership, then that is a valid use case for non-`const` reference.

If ownership is guaranteed to be transferred to the function, have the `unique_ptr` as a value parameter. It might require a `move` from the caller into the value, but it shows the caller that the object is not his concern anymore and the previous variable that held the pointer is guaranteed to be invalid after that. Both `unique_ptr` and `shared_ptr` are cheap to move.

If ownership transfer is not required, pass it as a `const` reference to the `unique_ptr`, if access to the smart pointer internals are required for some reason, or better yet, a raw pointer or reference to the object itself, as that will work irrespective of how the memory is being managed by the caller (smart pointers, new/delete, stack, etc...).

Reply:
That's a good point, thanks for noticing it! Fixed here: https://github.com/seladb/PcapPlusPlus/pull/1424/commits/17330d797dd5b398c82657b267b04fb032415f55